### ISSUE TITLE
Issues when using Picasso inside an IntentService with RemoteViews/Notifications

### ISSUE THREAD

**Issue by peitek** [square/picasso#1034]
Hi,

First of all, Picasso is a fantastic library and has been a pleasure to work with. I really like the great design and easy calls.

Lately, I have issues with an edge use of Picasso. I'm not sure if this is supposed to be supported, but I believe it is. It seems to me like a bug, but I might just be doing something wrong. Basically, the app is using [GCM](http://developer.android.com/google/gcm/index.html) to create notifications on the client phones. The notification uses custom views with RemoteViews. The RemoteView also has an ImageView. What I'd like to do is to load an image from the internet to that ImageView in the notification. The implementation is using an IntentService to handle the GCM handling in the background, which should be align with the recommended implementation. I can't really share the company code, so I created a simplified app, which shows all the issues I had in a (quickly duct-taped) minimal version of the app. You can find the repo [here](https://github.com/peitek/PicassoIntentServiceNotification).

I've tried three different things, which all had negative (side) effects. The entire code of the relevant class is in the [GcmIntentService](https://github.com/peitek/PicassoIntentServiceNotification/blob/master/app/src/main/java/io/futurestud/picassointentservicenotification/GcmIntentService.java).

## Attempt 1 - Result: Picasso is crashing

I created the notification with RemoteViews and some standard options:

```java
        mNotificationManager = (NotificationManager) this.getSystemService(Context.NOTIFICATION_SERVICE);

        RemoteViews remoteViews = new RemoteViews(getPackageName(), R.layout.remoteview_notification);
        remoteViews.setImageViewResource(R.id.remoteview_notification_icon, R.mipmap.ic_launcher);

        NotificationCompat.Builder mBuilder =
                new NotificationCompat.Builder(GcmIntentService.this)
                        .setSmallIcon(R.mipmap.ic_launcher)
                        .setContent(remoteViews)
                        .setPriority(NotificationCompat.PRIORITY_MAX)
                        .setAutoCancel(true);

        mBuilder.setDefaults(Notification.DEFAULT_VIBRATE);
        mBuilder.setDefaults(Notification.DEFAULT_SOUND);

        Notification notification = mBuilder.build();
        mNotificationManager.notify(notificationId, notification);
```

Next, I tried to load the image to the ImageView with Picasso. I used the suggested RemoteView `.into` implementation:

```java
        Picasso
                .with(GcmIntentService.this)
                .load("http://i.imgur.com/rFLNqWI.jpg")
                .into(remoteViews, R.id.remoteview_notification_icon, notificationId, notification);
```

This results in Picasso simply crashing. The thread in issue #547 explains that it's a threading issue. The suggested solution is to put it in a Handler/Runnable to run the Picasso call on the UI thread. 

## Attempt 2 - Notification sound plays twice

So, I followed the advice and wrapped the Picasso call in a Handler:

```java
Handler uiHandler = new Handler(Looper.getMainLooper());
        uiHandler.post(new Runnable() {
            @Override
            public void run() {
                Picasso
                        .with(GcmIntentService.this)
                        .load("http://i.imgur.com/rFLNqWI.jpg")
                        .into(remoteViews, R.id.remoteview_notification_icon, notificationId, notification);
            }
        });
```

Overall, this is as close as I got. This correctly loads the image. When the image is in cache, everything works perfectly. Unfortunately, when the image needs to be loaded via the network, it causes the notification sound/vibration to be played twice. Once, when the notification is initially shown, and another time when the image is loaded into the ImageView.

## Attempt 3 - Notification shows up twice

The original issue I had (besides the sound issue) was when I was working with notification tags (additional to notification ids) via the [notify(String tag, int id, Notification notification)](http://developer.android.com/reference/android/app/NotificationManager.html#notify(java.lang.String, int, android.app.Notification)) call. In this scenario the notification simply showed up twice. Once in the original version, and another time once the image was loaded.

```java
mNotificationManager.notify("notifyByTag", notificationId, notification);

        Handler uiHandler = new Handler(Looper.getMainLooper());
        uiHandler.post(new Runnable() {
            @Override
            public void run() {
                Picasso
                        .with(GcmIntentService.this)
                        .load("http://i.imgur.com/rFLNqWI.jpg")
                        .into(remoteViews, R.id.remoteview_notification_icon, notificationId, notification);
            }
        });
```

![double notification when using tags](https://cloud.githubusercontent.com/assets/3457989/7733944/edba1dae-ff32-11e4-8e55-051f6de9330b.png)

If I'm not missing anything, Picasso doesn't seem to support the tag/notification id combo, which is fine. But it'd be nice if I can get the RemoteView image working.

I've testet this on Android 4.4 and 5.1 and experienced the same result. If I'm doing something wrong, please let me know. I appreciate you taking the time to read and look over my issues!

Thanks you!
Norman

**Comment by peitek** [comment]
Of course, after posting I discover that a recent commit https://github.com/square/picasso/commit/b9541d7c119d0818f5cec6f96f51cdd5accc9e50 added notification tag support. My issues were based on Picasso 2.5.2. I'm going to try Picasso directly from the repo, but at least 2.5.2 still seems to have the double sound issue.

**Comment by mblcdr** [comment]
I confirm this issue. Notification sounds twice with Picasso 2.5.2. Attempt 2 (and the explanation) is exactly what is happening.

**Comment by adolfdsilva** [comment]
Tested on android 4.4, facing the same issue.

**Comment by jasonsaruulo** [comment]
What about using a target and showing the notification once the image was loaded ? Do you still got the issue with notification sound playing twice ? Works for me.

                        final Target target = new Target() {
                            @Override
                            public void onBitmapLoaded(
                                    Bitmap bitmap, Picasso.LoadedFrom loadedFrom) {
                                showNotification(bitmap);
                            }

                            @Override
                            public void onBitmapFailed(Drawable drawable) {
                                showNotification(null);
                            }

                            @Override
                            public void onPrepareLoad(Drawable drawable) {
                            }
                        };
                        Handler uiHandler = new Handler(Looper.getMainLooper());
                        uiHandler.post(new Runnable() {
                            @Override
                            public void run() {
                                Picasso.with(context)
                                    .load(imageUrl)
                                    .noFade()
                                    .into(target);
                            }
                        });

**Comment by marcfrost** [comment]
The problem with regards to using the target approach and only popping up the notification once the image has been downloaded is that the notifications are no longer "real time". They're subject to the image download duration. 

The delay may or may not be acceptable based on your application's requirements, but it's worth noting that it isn't ideal, especially when operating on something like an EDGE connection.

**Comment by guess** [comment]
Note also that with the Target approach, Picasso only keeps a weak reference to the `Target` object, so you need to keep a strong reference to it until it has finished loading, or else you might find that it was garbage collected before the Bitmap is returned, causing notifications to sometimes not appear.
